File: 2022/day11/solve-part-1.hs
import Data.Array
import Data.List
import Data.List.Split
--import System.IO.Unsafe  -- for debugging

-- Note: I assume that Int is 64 bits wide! If that doesn't hold, make sure to
-- replace Int with Integer (except maybe in the base-3 related functions)


-- Base 3 arithmetic functions. Lazy, so we only evaluate the digits we need.

toBase3 :: Int -> [Int]
toBase3 0 = []
toBase3 i = mod i 3:toBase3 (div i 3)

fromBase3 :: [Int] -> Int
fromBase3 [] = 0
fromBase3 (x:xs) = x + 3*(fromBase3 xs)

divBase3 :: [Int] -> (Int, [Int])
divBase3 [] = (0, [])
divBase3 (x:xs) = (x, xs)

addBase3 :: [Int] -> [Int] -> [Int]
addBase3 = go 0
  where
    go 0 [] [] = []
    go carry a b = mod sum 3 : go (div sum 3) aTail bTail
      where
        (aHead, aTail) = divBase3 a
        (bHead, bTail) = divBase3 b
        sum = aHead + bHead + carry

mulBase3 :: [Int] -> [Int] -> [Int]
mulBase3 [] _ = []
mulBase3 a  b = normalize 0 $ go [] b
  where
    go :: [[Int]] -> [Int] -> [Int]
    go []    []     = []
    go terms []     = let (sum, rest) = produceDigit terms in sum:go rest []
    go terms (b:bs) = let (sum, rest) = produceDigit (map (b*) a:terms) in sum:go rest bs

    produceDigit :: [[Int]] -> (Int, [[Int]])
    produceDigit nums = (sum (map head nums), filter (not.null) (map tail nums))

    normalize 0 [] = []
    normalize c [] = c `mod` 3:normalize (c `div` 3) []
    normalize c (x:xs) = let sum = x + c in sum `mod` 3:normalize (sum `div` 3) xs

-- Modulus m, multiplicative inverse of 3 in m
type Field = (Int, Int)

-- Value mod m, base-3 representation of the value
type Item = (Int, [Int])

-- I adapted this from: https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Pseudocode
modularInverse :: Int -> Int -> Int
modularInverse m a = calc (a, 1, 0, m, 0, 1)
  where
    step :: (Int, Int, Int, Int, Int, Int) -> (Int, Int, Int, Int, Int, Int)
    step (old_r, old_s, old_t, r, s, t) = (r, s, t, old_r - q * r, old_s - q * s, old_t - q * t)
      where q = old_r `div` r

    calc arg@(old_r, old_s, old_t, r, s, t)
      | r == 0    = if old_r == 1 then old_s else failed
      | otherwise = calc $ step arg
      where failed = error $ "Failed to calculate modular inverse of " ++ show a ++ " modulo " ++ show m ++ " (are arguments coprime?)"

makeField :: Int -> Field
makeField mod = (mod, modularInverse mod 3)

makeItem :: Field -> Int -> Item
makeItem (m, _) i = (i `mod` m, toBase3 i)

divides :: Item -> Int -> Bool
divides (a, _) b = a `mod` b == 0

addInt :: Field -> Item -> Int -> Item
addInt (m, _) (a, b) c = ((a + c) `mod` m, b `addBase3` toBase3 c)

addSelf :: Field -> Item -> Item
addSelf (m, _) (a, b) = ((a + a) `mod` m, addBase3 b b)

mulInt :: Field -> Item -> Int -> Item
mulInt (m, _) (a, b) c = ((a * c) `mod` m, b `mulBase3` toBase3 c)

mulSelf :: Field -> Item -> Item
mulSelf (m, _) (a, b) = ((a * a) `mod` m, b `mulBase3` b)

div3 :: Field -> Item -> Item
div3 (m, i3) (a, b) = (((a - head) * i3) `mod` m, tail) where (head, tail) = divBase3 b

data Monkey = Monkey
  { mItems  :: [Int]
  , mUpdate :: Field -> Item -> Item
  , mDiv    :: Int
  , mPos    :: Int
  , mNeg    :: Int
  };

parseUpdate :: String -> (Field -> Item -> Item)
parseUpdate (sOp:' ':sArg)
  = case sArg of
    "old" -> case sOp of
      '+' -> addSelf
      '*' -> mulSelf
    s -> let b = read s in case sOp of
      '+' -> \f a -> addInt f a b
      '*' -> \f a -> mulInt f a b

parseMonkey :: [String] -> Monkey
parseMonkey [_, line1, line2, line3, line4, line5]
  = Monkey { mItems=items, mUpdate=update, mDiv=divisor, mPos=pos, mNeg=neg }
  where
    Just sItems  = stripPrefix "  Starting items: " line1
    Just sUpdate = stripPrefix "  Operation: new = old " line2
    Just sDiv    = stripPrefix "  Test: divisible by " line3
    Just sPos    = stripPrefix "    If true: throw to monkey " line4
    Just sNeg    = stripPrefix "    If false: throw to monkey " line5
    items = map read $ splitOn ", " sItems
    update = \f i -> div3 f (((parseUpdate sUpdate) f) i)
    divisor = read sDiv
    pos = read sPos
    neg = read sNeg
parseMonkey lines = error $ "Unable to parse lines: " ++ show lines

parseInput :: String -> [Monkey]
parseInput = map parseMonkey . splitOn [""] . lines

numRounds = 20

solve :: [Monkey] -> Int
solve monkeyList = monkeyBusiness
  where
    m = foldl1 lcm $ map mDiv monkeyList
    f = makeField m

    numMonkeys = length monkeyList

    monkeys :: Array Int Monkey
    monkeys = listArray (0, numMonkeys - 1) monkeyList

    -- A list of pairs of initial <monkey index, item value> pairs
    monkeyItems :: [(Int, Item)]
    monkeyItems = [(index, makeItem f item) | (index, Monkey{mItems=items}) <- assocs monkeys, item <- items]

    itemPositions :: [(Int, Item)]
    itemPositions = concat $ map (takeRounds numRounds . iterate play) monkeyItems

    -- Given a <monkey index, item value> pair, updates the item value and
    -- calculates the next monkey index.
    play :: (Int, Item) -> (Int, Item)
    play (index, item) = (index', item')
      where
        Monkey{mUpdate=update, mDiv=divisor, mPos=pos, mNeg=neg} = monkeys ! index
        item'  = update f item
        index' = if item' `divides` divisor then pos else neg

    -- Truncates the infinite list of <monkey, item> pairs to the first given number of round.
    -- A new round starts whenever the next monkey index is less than the previous one.
    takeRounds :: Int -> [(Int, Item)] -> [(Int, Item)]
    takeRounds 0 _  = []
    takeRounds n (head@(i, _):tail@((j, _):_)) = head:takeRounds n' tail
      where n' = if i <= j then n else n - 1

    -- For each monkey, the number of items it has handled
    itemsHandled :: [Int]
    itemsHandled = elems a
      where
        a  = accumArray f 0 (0, numMonkeys - 1) itemPositions
        f cnt _ = cnt + 1

    -- The "monkey business" is the product of the two largest number of items handled by any monkeys
    monkeyBusiness = foldl1 (*) $ take 2 $ reverse $ sort itemsHandled

main = interact ((++ "\n") . show . solve . parseInput)

File: 2022/day11/solve-part-1.py
from math import lcm
from operator import add, mul
import re
import sys

class Base3:
  def __init__(self, value):
    assert isinstance(value, int)
    self.value = value % mod3k

  def DivideBy3(self):
    rem = self.value % 3
    self.value //= 3
    return rem

  def AddInt(self, i):
    self.value = (self.value + i) % mod3k

  def AddBase3(self, o):
    self.value = (self.value + o.value) % mod3k

  def MulInt(self, i):
    self.value = (self.value * i) % mod3k

  def MulBase3(self, o):
    self.value = (self.value * o.value) % mod3k

  def __str__(self):
    return str(self.value)



class Acc:
  def __init__(self, value, base3_value=None):
    if base3_value is None:
      base3_value = Base3(value)
    self.mod_value = value % mod
    self.base3_value = base3_value

  def ModAddDiv3(self, value):
    if isinstance(value, int):
      mod_value = (self.mod_value + value) % mod
      self.base3_value.AddInt(value)
    elif isinstance(value, Acc):
      assert value == self
      mod_value = (self.mod_value + value.mod_value) % mod
      self.base3_value.AddBase3(value.base3_value)
    else:
      assert False
    self.mod_value = ((mod_value - self.base3_value.DivideBy3()) * inv3) % mod
    return self

  def ModMulDiv3(self, value):
    if isinstance(value, int):
      mod_value = (self.mod_value * value) % mod
      self.base3_value.MulInt(value)
    elif isinstance(value, Acc):
      assert value == self
      mod_value = (self.mod_value * value.mod_value) % mod
      self.base3_value.MulBase3(value.base3_value)
    else:
      assert False
    self.mod_value = ((mod_value - self.base3_value.DivideBy3()) * inv3) % mod
    return self


  def Divides(self, div):
    #assert mod % div == 0
    return self.mod_value % div == 0

  def __str__(self):
    return "Acc{%s, %s}" % (self.mod_value, self.base3_value)

class MonkeyDef:
  '''Definition of a monkey. Constaints just the properties parsed from the
  input and is never modified afterwards. Can be used to initialize a Monkey.'''
  def __init__(self, items, update, div, pos, neg):
    self.items = items
    self.update = update
    self.div = div
    self.pos = pos
    self.neg = neg

class Monkey:
  '''Stateful monkey, that has a mutable list of items and knows how to play
  keep-away with other monkeys.'''
  def __init__(self, def_, monkeys):
    self.monkeys = monkeys

    # Start out with an initial list of items copied from the definition.
    self.items = list(map(Acc, def_.items))

    # Define the next-monkey function.
    div = def_.div
    pos = def_.pos
    neg = def_.neg
    self.nextMonkey = lambda i: pos if i.Divides(div) else neg

    # Define the item-update function which differs between part 1 and 2.
    self.update = def_.update

  def Play(self):
    for item in self.items:
      item = self.update(item)
      self.monkeys[self.nextMonkey(item)].items.append(item)
    self.items.clear()


OPS = {
  '+': lambda a, b: a.ModAddDiv3(b),
  '*': lambda a, b: a.ModMulDiv3(b),
}

def ParseUpdate(a, op, b):
  op = OPS[op]

  if a == 'old' and b == 'old':
    return lambda i: op(i, i)

  if a == 'old':
    b = int(b)
    return lambda i: op(i, b)

  assert False


def ParseMonkeyDef(index, lines):
  assert len(lines) == 6
  m0 = re.match('Monkey (\d+):', lines[0])
  m1 = re.match('  Starting items: (\d+(, \d+)*)', lines[1])
  m2 = re.match('  Operation: new = (old|\d+) ([+*]) (old|\d+)', lines[2])
  m3 = re.match('  Test: divisible by (\d+)', lines[3])
  m4 = re.match('    If true: throw to monkey (\d+)', lines[4])
  m5 = re.match('    If false: throw to monkey (\d+)', lines[5])
  assert index == int(m0[1])
  items = list(map(int, m1[1].split(', ')))
  update = ParseUpdate(m2[1], m2[2], m2[3])
  div = int(m3[1])
  pos = int(m4[1])
  neg = int(m5[1])
  assert pos != index and neg != index
  return MonkeyDef(items, update, div, pos, neg)


def PrintMonkeys(monkeys):
  for i, m in enumerate(monkeys):
    print('Monkey {}: {}'.format(i, ', '.join(map(str, m.items))))
  print()


def Solve(rounds):
  global mod3k
  mod3k = 3**(rounds  * max_moves_per_round)

  monkeys = []
  for def_ in monkey_defs:
    monkeys.append(Monkey(def_, monkeys))

  #PrintMonkeys(monkeys)
  inspected = [0]*len(monkeys)
  for r in range(rounds):
    # Lower modulo to account for fewer rounds remaining
    mod3k = 3**((rounds - r) * max_moves_per_round)

    for i, m in enumerate(monkeys):
      inspected[i] += len(m.items)
      m.Play()
    #PrintMonkeys(monkeys)

  return mul(*sorted(inspected)[-2:])


def Inverse(a):
  '''Computes the multiplicative inverse of `a` modulo `mod`, assuming `a` and `mod` are coprime.'''
  # Extended Euclidean algorithm
  old_r, r = a, mod
  old_s, s = 1, 0
  old_t, t = 0, 1
  while r != 0:
    q = old_r // r
    old_r, r = r, old_r - q * r
    old_s, s = s, old_s - q * s
    old_t, t = t, old_t - q * t
  assert old_r == 1
  return old_s

monkey_defs = [ParseMonkeyDef(i, part.splitlines())
    for i, part in enumerate(sys.stdin.read().split('\n\n'))]

mod = lcm(*(m.div for m in monkey_defs))
assert mod % 3 != 0
inv3 = Inverse(3)
assert (inv3 * 3) % mod == 1

# Maximum number of times an item can move in a single round.
max_moves = [None]*len(monkey_defs)
for i, m in reversed(list(enumerate(monkey_defs))):
  max_moves[i] = max([1] + [max_moves[j] + 1 for j in [m.pos, m.neg] if j > i])
max_moves_per_round = max(max_moves)

print(Solve(20))
